solver/src/settlement_submission.rs

```diff
@@ -55,7 +66,25 @@ pub async fn submit(
     };
     // We never cancel.
     let cancel_future = std::future::pending::<CancelSender>();
-    let stream = gas_price_stream(target_confirm_time, gas_price_cap, gas_limit, gas);
+    let mut stream = gas_price_stream(target_confirm_time, gas_price_cap, gas_limit, gas).boxed();
+
+    // It is possible that there is a pending transaction we don't know about because the driver
+    // got restarted while it was in progress. Sending a new transaction could fail in that case
+    // because the gas price has not increased. So we make sure that the starting gas price is at
+    // least high enough to accommodate. This isn't perfect because it's still possible that that
+    // transaction gets mined first in which case our new transaction would fail with "nonce already
+    // used".
+    if let Some(gas_price) = pending_gas_price {
+        tracing::info!(
+            "detected existing pending transaction with gas price {}",
+            gas_price
+        );
+        let gas_price = gas_price.to_f64_lossy();
+        let gas_price = transaction_retry::gas_price_increase::minimum_increase(gas_price);
+        stream = futures::stream::once(std::future::ready(gas_price))
```

Comment:
Isn't retry expecting a monotonically increasing gas price stream? Here is looks as if it was possible to have the previous (potentially high value, followed by a lower value).

Maybe this would be a good moment to refactor `gas_price_stream` to a proper struct which can be composed like:

```rust
let stream = GasPriceStream::new(target_confirm_time, gas_limit, estimator)
  .capped(gas_price_cap);

if let Some(gas_price) = pending_gas_price {
  stream = stream.previous_value(gas_price)
}

stream = stream.minimum_increase();
```

Of course this can be done as functions as well.

Reply:
Oh right, for some reason I thought the enforce_min_increase adapter would handle this but obviously it is only applied to the second stream. I have changed `gas_price_stream` to take another parameter which passes an optional intial gas price. Will think about refactoring it more now.